core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4557,31 +4560,43 @@ private QueryModel rewriteSampleBy(@Nullable QueryModel model) throws SqlExcepti
-            ObjList<ExpressionNode> sampleByFill = nested.getSampleByFill();
-            ExpressionNode sampleByTimezoneName = nested.getSampleByTimezoneName();
-            ExpressionNode sampleByUnit = nested.getSampleByUnit();
-            ExpressionNode timestamp = nested.getTimestamp();
+            final ExpressionNode sampleBy = nested.getSampleBy();
+            final ExpressionNode sampleByOffset = nested.getSampleByOffset();
+            final ObjList<ExpressionNode> sampleByFill = nested.getSampleByFill();
+            final ExpressionNode sampleByTimezoneName = nested.getSampleByTimezoneName();
+            final ExpressionNode sampleByUnit = nested.getSampleByUnit();
+            final ExpressionNode timestamp = nested.getTimestamp();
+            final int sampleByFillSize = sampleByFill.size();
+            ExpressionNode sampleByFrom = nested.getSampleByFrom();
+            ExpressionNode sampleByTo = nested.getSampleByTo();
 
             if (
                     sampleBy != null
                             && timestamp != null
                             && (sampleByOffset != null && SqlKeywords.isZeroOffset(sampleByOffset.token) && (sampleByTimezoneName == null || SqlKeywords.isUTC(sampleByTimezoneName.token)))
-                            && (sampleByFill.size() == 0 || (sampleByFill.size() == 1 && SqlKeywords.isNoneKeyword(sampleByFill.getQuick(0).token)))
+                            && (sampleByFillSize == 0 || (sampleByFillSize == 1 && !SqlKeywords.isPrevKeyword(sampleByFill.getQuick(0).token) && !SqlKeywords.isLinearKeyword(sampleByFill.getQuick(0).token)))
                             && sampleByUnit == null
+                            && !((sampleByFrom != null && sampleByFrom.token.charAt(0) == '$') || (sampleByTo != null && sampleByTo.token.charAt(0) == '$'))
             ) {
                 // Validate that the model does not have wildcard column names.
                 // Using wildcard in group-by expression makes SQL ambiguous and
                 // error-prone. For example, additive table metadata changes (adding a column)
                 // will change the outcome of existing queries, if those supported wildcards for
                 // as group-by keys.
 
+                ObjList<ExpressionNode> maybeKeyed = new ObjList<>();
+
                 for (int i = 0, n = model.getColumns().size(); i < n; i++) {
-                    QueryColumn column = model.getColumns().getQuick(i);
-                    if (column.getAst().isWildcard()) {
+                    final QueryColumn column = model.getColumns().getQuick(i);
+                    final ExpressionNode ast = column.getAst();
+                    if (ast.isWildcard()) {
                         throw SqlException.$(column.getAst().position, "wildcard column select is not allowed in sample-by queries");
                     }
+
+                    // it might be a keyed expression
+                    if (ast.type == LITERAL) {
```

Comment:
Keys may also be functions and expressions. Don't we need to handle these cases?